sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -187,14 +202,29 @@ object GpuCast {
 
   val INVALID_NUMBER_MSG: String = "At least one value is either null or is an invalid number"
 
+  private def ansiMode = SQLConf.get.getConf(SQLConf.ANSI_ENABLED)
+
+  private def legacyCastString = SQLConf.get.getConf(SQLConf.LEGACY_COMPLEX_TYPES_TO_STRING)
```

Comment:
These variables will only get initialized once, using the config at that time. The config could change between different queries being executed. The `doCastr` method already has `ansiMode` and `legacyCastToString` passed in, so I'm not sure why we need them here as well.

Reply:
There are instance variables that are dependent on the conf, they are outside of the doCast function. 

These `variables` are functions and will be updated every time an instance is made and thus will pick up the new conf. Am I missing something?

Reply:
I did miss that these are `def` not `val`, but it seems odd to have them defined both as variables at the object level and also have them passed in to the `doCast` method.

Reply:
If I do an `ansi_cast` instead of a `cast`, Does that mean that `useDecimalPlainString` will be wrong? Or what about if I change  `spark.sql.ansi.enabled` part way through a session?  It is perfectly valid to change it. Would that impact the result of `useDecimalPlainString` and possible return the wrong result for a decimal value?  `legacyCastToString is similar. I can change it at runtime.

Reply:
Sorry I missed the part about them being functions. The big issue is then `ansi_cast`, where I can have a Cast expression that is marked as ANSI, but the config is not set for ANSI.

Reply:
I didn't realize there was a way to invoke ANSI without setting the conf. If that is the case we definitely have a problem. Can you please give me an example? The only way I could come up with is to write a hive table and insert decimal values into a string column.

Reply:
Sorry `ansi_cast` is not a real operator. The point is that the API provides one set of expectations, but does not live up to them.  That for me is a bug.  We have our own code that violates this constraint.

https://github.com/NVIDIA/spark-rapids/blob/01fce6c909cae55bd1b8146cc63e9ec055316ad7/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala#L384

What is to stop spark from doing it in the future?  It makes the code brittle.

Reply:
I agree, I am not too happy with this situation either now that you mention it. I designed it to work with the `CastOperation` solution but now I am just trying to contort this into position. I need to take a step back and think this through